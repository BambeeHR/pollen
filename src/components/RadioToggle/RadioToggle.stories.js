import RadioToggle from './RadioToggle.vue';
import '../../assets/tailwind.css';

export default {
  title: 'RadioToggle',
  component: RadioToggle,
  story: {
    parameters: {
      jest: ['RadioToggle.test.js'],
    },
  },
};

export const Default = () => ({
  components: { RadioToggle },
  data() {
    return {
      options: [
        'Option 1',
        {
          label: 'Option 2',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: '',
    };
  },
  template: '<RadioToggle v-model="value" :options="options" />',
});

export const WithInitialValue = () => ({
  components: { RadioToggle },
  data() {
    return {
      options: [
        'Option 1',
        {
          label: 'Option 2',
          value: 'foo',
        },
        {
          label: 'Option 3',
          value: 'bar',
        },
      ],
      value: 'foo',
    };
  },
  template: '<RadioToggle v-model="value" :options="options" />',
});

/* eslint-disable no-param-reassign */
[Default, WithInitialValue].forEach((item) => {
  item.story = {
    parameters: {
      jest: ['RadioToggle.test.js'],
    },
  };
});
