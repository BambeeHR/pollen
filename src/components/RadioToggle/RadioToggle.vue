<template>
  <div
    class="radio-toggle"
    :class="{ focused: isFocused }"
    @click="isFocused = false"
  >
    <label
      v-for="(item, i) in mappedOptions"
      :key="item.value"
      :ref="item.value"
      :for="`${name}-${i}`"
      :class="{ active: value === item.value }"
      class="radio-toggle-option"
    >
      <input
        :id="`${name}-${i}`"
        :value="item.value"
        :name="name"
        type="radio"
        class="radio-toggle-input"
        @blur="isFocused = false"
        @focus="isFocused = true"
        @input="handleInput"
      />
      <span class="radio-toggle-label">{{ item.label }}</span>
    </label>
    <div
      v-if="refsMounted && value"
      class="radio-toggle-indicator"
      :style="bgStyle"
    />
  </div>
</template>

<script>
import shortid from 'shortid';

export default {
  props: {
    value: {
      type: String,
      default: null,
    },
    options: {
      type: Array,
      default: () => [],
    },
    name: {
      type: String,
      default: shortid.generate,
    },
  },

  data() {
    return {
      isFocused: false,
      refsMounted: false,
    };
  },

  computed: {
    mappedOptions() {
      return this.options.map((item) => {
        if (typeof item === 'string') {
          return {
            label: item,
            value: item,
          };
        }
        const { label, value } = item;

        return {
          label,
          value,
        };
      });
    },

    bgStyle() {
      const activeLabel = this.$refs[this.value]
        ? this.$refs[this.value][0]
        : null;
      if (!activeLabel) {
        return {};
      }

      return {
        left: `${activeLabel.offsetLeft}px`,
        height: `${activeLabel.offsetHeight}px`,
        width: `${activeLabel.offsetWidth}px`,
      };
    },
  },

  mounted() {
    this.$nextTick(() => {
      this.refsMounted = true;
    });
  },

  methods: {
    handleInput(e) {
      this.$emit('input', e.target.value);
    },
  },
};
</script>

<style scoped>
.radio-toggle {
  @apply bg-gray-100
    border 
    border-grey-200
    border-solid
    inline-flex
    p-1 
    relative
    rounded-full;
}

.radio-toggle-option {
  @apply cursor-pointer
    duration-100
    ease-linear
    font-bold 
    px-5 
    py-1
    relative 
    rounded-full
    select-none
    text-center 
    text-2xs 
    tracking-wide 
    transition-all
    uppercase 
    z-10;
}

.radio-toggle-option.active {
  @apply text-white;
}

.radio-toggle.focused {
  @apply shadow-outline;
}

.radio-toggle-input {
  @apply sr-only;
}

.radio-toggle-indicator {
  @apply absolute 
    bg-purple 
    duration-100
    ease-in
    rounded-full 
    transition-all
    z-0;
}
</style>
