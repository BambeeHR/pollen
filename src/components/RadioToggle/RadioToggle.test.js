import { mount } from '@vue/test-utils';
import RadioToggle from './RadioToggle.vue';

const options = [
  'Option 1',
  {
    label: 'Option 2',
    value: 'foo',
  },
  {
    label: 'Option 3',
    value: 'bar',
  },
];

describe('RadioToggle', () => {
  test('renders correctly', () => {
    const wrapper = mount(RadioToggle, {
      propsData: {
        options,
        value: 'foo',
        name: 'test',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('highlights the correct label', () => {
    const wrapper = mount(RadioToggle, {
      propsData: {
        options,
        value: 'foo',
      },
    });
    expect(wrapper.find('.active').text()).toBe('Option 2');
  });

  test('emits an input event', async () => {
    const wrapper = mount(RadioToggle, {
      propsData: {
        options,
      },
    });
    const inputs = wrapper.findAll('input');
    expect(inputs.length).toBe(3);
    await inputs.at(1).trigger('input');
    await inputs.at(0).trigger('input');
    await inputs.at(2).trigger('input');
    expect(wrapper.emitted().input).toEqual([['foo'], ['Option 1'], ['bar']]);
  });
});
