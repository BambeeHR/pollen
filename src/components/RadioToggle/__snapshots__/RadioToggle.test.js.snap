// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioToggle renders correctly 1`] = `
<div
  class="radio-toggle"
>
  <label
    class="radio-toggle-option"
    for="test-0"
  >
    <input
      class="radio-toggle-input"
      id="test-0"
      name="test"
      type="radio"
      value="Option 1"
    />
     
    <span
      class="radio-toggle-label"
    >
      Option 1
    </span>
  </label>
  <label
    class="radio-toggle-option active"
    for="test-1"
  >
    <input
      class="radio-toggle-input"
      id="test-1"
      name="test"
      type="radio"
      value="foo"
    />
     
    <span
      class="radio-toggle-label"
    >
      Option 2
    </span>
  </label>
  <label
    class="radio-toggle-option"
    for="test-2"
  >
    <input
      class="radio-toggle-input"
      id="test-2"
      name="test"
      type="radio"
      value="bar"
    />
     
    <span
      class="radio-toggle-label"
    >
      Option 3
    </span>
  </label>
   
  <!---->
</div>
`;
